### ISSUE TITLE
What is hier_thresh? What is nms_thresh?

### ISSUE THREAD

**Issue by LukeAI** [AlexeyAB/darknet#3887]
In get_network_boxes, there is a parameter called "hier_thresh" - please could you explain what this is - varying from 0.1 to 0.9 doesn't seem to affect my detections.

I know that NMS is for trimming down thousands of anchor boxes probably detecting the same object to just one - but what exactly is the role of nms_thresh in that?

**Comment by AlexeyAB** [comment]
hier_thresh - is for Yolo9000 (hierarchy threshold)

nms_thresh - if IoU > IoU_threshold, then these 2 boxes will be fused into 1.

**Comment by anisaha1** [comment]
Hi @AlexeyAB, I have some questions about evaluation thresholds. There are a lot of parameters involved and I when I compare the thresholds in different implementations of YOLO, I find inconsistencies. It would be great if you could help me.

I understand ```thresh_calc_avg_iou = 0.25``` and ```iou_thresh = 0.5``` which is consistent in all the functions. The first is the probability confidence threshold and the second the IoU threshold.
 
There are two other thresholds which are passed as arguments to the get_network_boxes function ```thresh``` and ```hier_thresh```.
thresh is set to ```const float thresh = 0.005;``` in validate_detector_map
[https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L944](url)

it is set to 0.001 in validate_detector
[https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L681](url)
0.001 in validate_detector recall
[https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L815](url)

What is the purpose of thresh?

I understand as you mentioned above that hier_thresh is for YOLO9000 which does joint classification and detection and creates a wordnet tree. Does YOLOv3 also use this technique?

Why is hier_thresh set to 0 in train_detector,
[https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L1007](url)

but to 0.5 on validation?
[https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L1890](url)

Also about nms,
it's 0.45 here [https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L945](url)
but 0.4 here [https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L817](url)

When I was going through a Pytorch implementation of YOLO, the thresholds used are different. For example, the default nms threshold is set to 0.6 here [https://github.com/ultralytics/yolov3/blob/master/test.py#L11](url) I was wondering wouldn't the mAP evaluations vary because of the variation in thresholds? In such case, how do you make model mAP comparisons consistent across implementations?

**Comment by AlexeyAB** [comment]
@anisaha1 

It seems that URLs in your message are broken.

> it is set to 0.001 in validate_detector
> https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L681
 
Only this value (and function) is used for detections for AP calculation on MSCOCO test-dev.
`./darknet detector valid ...`


> I understand thresh_calc_avg_iou = 0.25 and iou_thresh = 0.5 which is consistent in all the functions. The first is the probability confidence threshold and the second the IoU threshold.
> 
> There are two other thresholds which are passed as arguments to the get_network_boxes function thresh and hier_thresh.
> thresh is set to const float thresh = 0.005; in validate_detector_map

For our internal AP calculation we use ./darknet detector map ...
`const float thresh = 0.005;` is used for our internal AP calculation.
While `thresh_calc_avg_iou = 0.25` is used for average IoU calculation only.

> I was wondering wouldn't the mAP evaluations vary because of the variation in thresholds? 

AP are calculated **for all possible confidence_thresholds** higher than the minimal value.

> In such case, how do you make model mAP comparisons consistent across implementations?

Published AP are calculated for the test-dev dataset **without labels** on the evaluation server. So the MS COCO server do comparison for test-dev dataset.

All parameters affect accuracy. Many paper/ detectors use various tricks: test time data augmentation, ensembles, training on 100 GPUs with large mini-batch size, .... - we don't use these tricks.

**Comment by anisaha1** [comment]
Hi @AlexeyAB, thanks for your quick response. I had messed up the URL positions. They are now fixed.

Your explanation answers some of my questions. I still have a few more.
I now understand that only ```./darknet detector valid ...``` is used for MSCOCO test-dev validation before you send it to the evaluation server. AP is calculated for all possible conf_thresholds higher than minimum value. Is this minimum value standard (like 0.005)? For example when I compare the functions below, one uses 0.001 but another 0.005. 

[validate_detector AlexeyAB](https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L681) vs [validate_detector pjreddie](https://github.com/pjreddie/darknet/blob/master/examples/detector.c#L419)

Essentially, I guess what I am asking is does everybody use the minimum conf threshold (0.001/0.005) and nms threshold ([0.45](https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L945)/[0.6](https://github.com/ultralytics/yolov3/blob/master/test.py#L11))  at which their model performs best? Shouldn't every evaluation use the same thresholds?

Also when you say "AP are calculated for all possible confidence_thresholds higher than the minimal value", does that mean you evaluate AP internally on COCOval on different conf thresholds increasing from [0.001, 0.005 ... ] and choose the threshold which suits you best (TP, FP values etc.)?

I still don't understand why hier_thresh is set to 0.5 in [validate_detector](https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L728) but to 0 in [validate_detector_map](https://github.com/AlexeyAB/darknet/blob/master/src/detector.c#L1007)? Why is your internal mAP calculation hierarchy threshold different from what you use when you send to MSCOCO test-dev evaluation server?

**Comment by AlexeyAB** [comment]
> Is this minimum value standard (like 0.005)?

This is a hyperparameter of model, like many others in cfg-file. Just hyperparameters `thresh` and `nms_thresh` are hardcoded in the source code in the Darknet. Each model has its own optimal value, so it should be tuned for each model.

> validate_detector AlexeyAB vs validate_detector pjreddie

Joseph used 0.001 before https://github.com/pjreddie/darknet/blame/9a01e6ccb7a74ff77e99060cf18acd6cfdb74b8e/src/detector.c#L233

> Shouldn't every evaluation use the same thresholds?

No. 
For each model, its own is optimal, and each model uses different: weights, anchors, thresholds, ...
MS COCO suggests that you can use any weights, hyperparameters and thresholds.

>  Why is your internal mAP calculation hierarchy threshold different from what you use when you send to MSCOCO test-dev evaluation server?

hier_thresh is used for Yolo9000 only.
It isn't used for Yolo v1-v4 and isn't used for MS COCO training/testing.

**Comment by anisaha1** [comment]
Great! Thanks a lot! That clears things up. Also, congratulations on your YOLOv4 paper on arXiv.
